CDP4WebServices.API/Services/BusinessLogic/Implementation/ChainOfRdlComputationService.cs

```diff
@@ -0,0 +1,187 @@
+                    continue;
+                }
+
+                var modelReferenceDataLibararyIid = engineeringModelSetup.RequiredRdl.SingleOrDefault();
+                if (modelReferenceDataLibararyIid == Guid.Empty)
+                {
+                    Logger.Warn($"The EngineeringModelSetup { engineeringModelSetup.Iid } has more than 1 required rdl, this is not allowed, this EngineeringModelSetup is ignored");
+                    continue;
+                }
+
+                var modelReferenceDataLibarary = this.cachedModelReferenceDataLibraries.SingleOrDefault(x => x.Iid == modelReferenceDataLibararyIid);
+                if (modelReferenceDataLibarary == null)
+                {
+                    Logger.Warn($"The ModelReferenceDataLibarary { modelReferenceDataLibararyIid } could not be found, there is a fault in the data, the EngineeringModelSetup {engineeringModelSetup.Iid} is ignored");
+                    continue;
+                }
+                else
+                {
+                    this.QueryRequiredReferenceDataLibraryChainAndUpdateResult(modelReferenceDataLibarary, result);
+                }
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// recursively queries the chain of rdls for a provided <see cref="ReferenceDataLibrary"/> and adds the found <see cref="SiteReferenceDataLibrary"/> to the
+        /// resulting <paramref name="siteReferenceDataLibraryUniqueIdentifiers"/>
+        /// </summary>
+        /// <param name="referenceDataLibrary">
+        /// The <see cref="ReferenceDataLibrary"/> for which the chain-of-rdl is to be computed
+        /// </param>
+        /// <param name="siteReferenceDataLibraryUniqueIdentifiers">
+        /// A <see cref="HashSet{Guid}"/> to which the 
+        /// </param>
+        private void QueryRequiredReferenceDataLibraryChainAndUpdateResult(ReferenceDataLibrary referenceDataLibrary, HashSet<Guid> siteReferenceDataLibraryUniqueIdentifiers)
+        {
+            // the referenceDataLibrary does not have a requiredRdl, no need to continue
+            if (!referenceDataLibrary.RequiredRdl.HasValue)
+            {
+                return;
+            }
+
+            // the required RDL of the referenceDataLibrary has already been found in a previous iteration, no need to redo it
+            if (siteReferenceDataLibraryUniqueIdentifiers.Contains(referenceDataLibrary.RequiredRdl.Value))
+            {
+                return;
+            }
+            
+            // iterate through the cached SiteReferenceDataLibrary objects untill we find a matching required RDL.
+            // Add this matched object to the resultset and look for it's required rdl's with a recursive call
+            foreach (var siteReferenceDataLibrary in this.cachedSiteReferenceDataLibraries)
```

Comment:
Just because you're not using Linq I started wondering why not.

I would expect a Linq query with an if () instead of a foreach () loop.
Something like:  
var requiredRdl = this.cachedSiteReferenceDataLibraries.SingleOrDefault(x => x.Iid == referenceDataLibrary.RequiredRdl)
if (requiredRdl != null) { ........... }

Or maybe Where(), instead of SingleOrDefault(). Depends on the level of error handling you want to implement.